lib/puppet/type/file.rb

```diff
@@ -641,12 +641,33 @@ def recurse_local
 
   # Recurse against our remote file.
   def recurse_remote(children)
+    recurse_remote_metadata do |meta|
+      if meta.relative_path == "."
+        parameter(:source).metadata = meta
+        next
+      end
+      children[meta.relative_path] ||= newchild(meta.relative_path)
+      children[meta.relative_path][:source] = meta.source
+      children[meta.relative_path].parameter(:source).metadata = meta
+    end
+
+    children
+  end
+
+  def recurse_remote_metadata(&block)
     sourceselect = self[:sourceselect]
 
     total = self[:source].collect do |source|
       next unless result = perform_recursion(source)
       return if top = result.find { |r| r.relative_path == "." } and top.ftype != "directory"
-      result.each { |data| data.source = "#{source}/#{data.relative_path}" }
+      result.each do |data|
+        if data.relative_path == '.'
+          data.source = source
+        else
+          # REMIND: appending file paths to URL may not be safe, e.g. foo+bar
+          data.source = "#{source}/#{data.relative_path}"
+        end
```

Comment:
Could you split out the refactor of `recurse_remote` so it calls `recurse_remote_metadata` into one commit and preserve the old behavior of:

```
result.each { |data| data.source = "#{source}/#{data.relative_path}" }
```

And then add a second commit for the compiler to call the new `recurse_remote_metadata` method and makes this change (so `data.source = source` when relative_path is `.`?) That way the refactor doesn't make any behavior changes?